### PR TITLE
Lab1 solution

### PR DESCRIPTION
**Pull request by mryab**

### CODE REVIEW

**Comment by mryab** [comment]
test

File: L1 - Gradient descent and linear models.py [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+
+
+def logistic_loss_m(x): return np.log(1 + np.exp(-x))
+
+
+plot_setup('M', 'Q', True, True, True)
+plt.plot(m, list(map(exact_loss_m, m)), label='$[M<0$]')
+plt.plot(m, list(map(mse_loss_m, m)), label='$(M - 1)^2$')
+plt.plot(m, list(map(hinge_loss_m, m)), label='$max(0,1-M)$')
+plt.plot(m, list(map(logistic_loss_m, m)), label='$\ln(1+e^{-M})$')
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: test

File: L1 - Gradient descent and linear models.py [hunk 2]
```diff
@@ -0,0 +1,690 @@
...
+first = np.random.normal(-1, 1, (NUM_SAMPLES // 2, 2))
+second = np.random.normal(1.7, 1, (NUM_SAMPLES // 2, 2))
+
+X = np.c_[np.ones(NUM_SAMPLES), np.concatenate((first, second))]
+Y = np.r_[np.ones(NUM_SAMPLES // 2), -np.ones(NUM_SAMPLES // 2)].reshape(NUM_SAMPLES)
+
+w_exact = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
+
+x_ax = np.linspace(-5, 5, 2)
+y_ax_exact = np.array(w_exact[0] / w_exact[1] - (w_exact[2] / w_exact[1]) * x_ax)
```
Comment (shirinnikita) [inline L1 - Gradient descent and linear models.py]: Да ладно тебе, на практике квадратичная ошибка оказалась не такой уж и плохой, как показалось на первый взгляд, и не такой плохой, как нарисовано у тебя - первое слагаемое тоже должно идти с отрицательным знаком, и тогда разделяющая прямая будет действительно разделять облака

File: L1 - Gradient descent and linear models.py [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+import math
+
+get_ipython().magic('matplotlib notebook')
+matplotlib.rcParams['figure.figsize'] = '12,8'
+matplotlib.rcParams['figure.max_open_warning'] = False
+
+
+# In[2]:
+
+def plot_figure_setup(xlabel='x', ylabel='y', hline=False, vline=False, equal_axes=False):
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Обычно имя функции начинают с _правильного_ глагола, к примеру, `setup_figure`

File: L1 - Gradient descent and linear models.py [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+    if vline:
+        plt.axvline(color='k', alpha=0.7)
+    return f
+
+
+# In[3]:
+
+SAMPLE_NUM = 20000
+
+first = np.random.normal(-1, 1, (SAMPLE_NUM // 2, 2))
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Не самое говорящее название класса, обычно пишут _positive_class_features_ или _negative_class_features_, для бинарной классификации. Или в нашем случае еще можно _red_class_features_ и _green_class_features_.

А еще проще не делить, а просто хранить количество элементов для одного класса ;)

File: L1 - Gradient descent and linear models.py [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+second = np.random.normal(1.7, 1, (SAMPLE_NUM // 2, 2))
+
+X = np.c_[np.ones(SAMPLE_NUM), np.concatenate((first, second))]
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Есть удобные функции `vstack` и `hstack`.

File: L1 - Gradient descent and linear models.py [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+Y = np.r_[np.ones(SAMPLE_NUM // 2), -np.ones(SAMPLE_NUM // 2)].reshape(SAMPLE_NUM)
+
+w_exact = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: А чем плохо простое умножение матриц? зачем dot?

File: L1 - Gradient descent and linear models.py [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+
+x_ax = np.linspace(-5, 5, 2)
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Очень странно, почему бы просто не взять `x=-5`, а потом `x=5`.

File: L1 - Gradient descent and linear models.py [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+y_ax_exact = np.array(-(w_exact[0] / w_exact[1]) - (w_exact[2] / w_exact[1]) * x_ax)
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: На самом деле у тебя
`w[0] + w[1]x + w[2]y = 0`, а это значит, что  `y = -(w[0] + w[1]x)/w[2]`,
 кстати `w[2]` неплохо бы проверять на ноль, но для лабы пойдет :)

File: L1 - Gradient descent and linear models.py [hunk 9]
```diff
@@ -0,0 +1,701 @@
...
+
+plot_figure_setup(hline=True, vline=True, equal_axes=True)
+plt.scatter(first.T[0], first.T[1], c='green', alpha=0.2)
+plt.scatter(second.T[0], second.T[1], c='red', alpha=0.2)
+plt.plot(x_ax, y_ax_exact.T, color='b')
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Нет причин для транспонирования.

File: L1 - Gradient descent and linear models.py [hunk 10]
```diff
@@ -0,0 +1,701 @@
...
+X = np.c_[np.ones(SAMPLE_NUM), np.concatenate((first, second))]
+Y = np.r_[np.ones(SAMPLE_NUM // 2), -np.ones(SAMPLE_NUM // 2)].reshape(SAMPLE_NUM)
+
+w_exact = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(Y)
+
+x_ax = np.linspace(-5, 5, 2)
+y_ax_exact = np.array(-(w_exact[0] / w_exact[1]) - (w_exact[2] / w_exact[1]) * x_ax)
+
+plot_figure_setup(hline=True, vline=True, equal_axes=True)
+plt.scatter(first.T[0], first.T[1], c='green', alpha=0.2)
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Можно писать так `plt.scatter(X[1, :], X[2, :])`

File: L1 - Gradient descent and linear models.py [hunk 11]
```diff
@@ -0,0 +1,701 @@
...
+# есть не что иное, как максимизация правдоподобия.
+# 2. Отобразите все функционалы качества в осях $M \times Q$ для одного элемента.
+
+# 
+# 1. Функция правдоподобия: $\mathbb{P}\{y|x\}=\sigma(y(wx+b))$. Если считать совпадения предсказанных классов для
+# элементов выборки с актуальными классами независимыми в совокупности событиями, то можно переписать $\mathbb{P}\{
+# y|x\}$ как $\prod_{i=0}^n\mathbb{P}\{y_i|x_i\}$. Максимизация функции правдоподобия тогда представляется как задача
+#  $$\underset{w}{\mathrm{argmax}}\prod_{i=0}^n\mathbb{P}\{y_i|x_i\}=\underset{w}{\mathrm{argmax}}\ln\prod_{
+# i=0}^n\mathbb{P}\{y_i|x_i\}=$$ (исходя из монотонности логарифма, если функция в данной точке имеет максимум,
+# то и её логарифм)$$=\underset{w}{\mathrm{argmax}}\sum_{i=0}^n\ln\mathbb{P}\{y_i|x_i\}=\underset{w}{\mathrm{
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Не самый очевидный переход, мы же только знаем, что  P(y=1|x) = sigma(wx), а ты подставляешь выражение с $sigma(ywx)$

File: L1 - Gradient descent and linear models.py [hunk 12]
```diff
@@ -0,0 +1,701 @@
...
+
+
+def logistic_loss_m(x): return np.log(1 + np.exp(-x))
+
+
+plot_figure_setup('M', 'Q', True, True, True)
+plt.plot(margin, list(map(exact_loss_m, margin)), label='$[M<0$]')
+plt.plot(margin, list(map(mse_loss_m, margin)), label='$(M - 1)^2$')
+plt.plot(margin, list(map(hinge_loss_m, margin)), label='$max(0,1-M)$')
+plt.plot(margin, list(map(logistic_loss_m, margin)), label='$\ln(1+e^{-M})$')
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Тут все классно, только тут можно было записать короче, к примеру, `np.log(1 + np.exp(x))`, все операции над вектором `x` буду поэлементные и кода будет поменьше, его прям по месту здесь можно было написать.

File: L1 - Gradient descent and linear models.py [hunk 13]
```diff
@@ -0,0 +1,701 @@
...
+#     3) $Q = max(0, 1 - y\cdot(wx))=max(0,1-M)$
+#     
+#     4) $Q = \ln(1 + e^{-y\cdot(wx)})=\ln(1+e^{-M})$
+
+# In[4]:
+
+margin = np.linspace(-7, 7, 1000)
+
+
+def exact_loss_m(x): return 1 if x < 0 else 0
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Не экономь переносы.

File: L1 - Gradient descent and linear models.py [hunk 14]
```diff
@@ -0,0 +1,701 @@
...
+             calculate_acc(sgd(start, trainset, labels, logistic_loss, logistic_grad, 1e-6, 1, b_size)[0], testset,
+                           testset_labels)[1], 'bo')
+plt.show()
+
+# Заметим, что размер батчей не влияет на качество классификации заметным образом за исключением случая,
+# когда выборка делится на два батча только с нулями и единицами (размер каждого - примерно 2000)
+
+# In[31]:
+
+plot_figure_setup('Batch size', 'Loss')
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: повторюсь, что подобные вещи рисуют для батчей.

File: L1 - Gradient descent and linear models.py [hunk 15]
```diff
@@ -0,0 +1,701 @@
...
+
+xx = np.arange(-10, 10, 0.01)
+yy = np.arange(-10, 10, 0.01)
+xgrid, ygrid = np.meshgrid(xx, yy)
+zgrid = simple1_func((xgrid, ygrid))
+plot_figure_setup()
+cont = plt.contour(xgrid, ygrid, zgrid)
+cont.clabel(fmt="%.0f")
+
+start_simple_1 = np.random.randn(2) * 10
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: ну мешанина же :) надо как-то получше форматировать + такие вещи обычно в цикле все же делают ;)

File: L1 - Gradient descent and linear models.py [hunk 16]
```diff
@@ -0,0 +1,701 @@
...
+ax.plot(xs=res_nes[2].T[0], ys=res_nes[2].T[1], zs=rosenbrock(res_nes[2].T), label='Nesterov\'s accelerated gradient')
+plt.legend()
+fig.show()
+
+
+# **Adagrad (2011)**
+
+# In[55]:
+
+def adagrad(w, func, grad, learn_rate, iter_num):
```
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Вот, все что ниже достаточно медленно сходится по целевой функции, хоть и качество неплохое. Думаю, что ты с learning_rate переусердствовал )

File: L1 - Gradient descent and linear models.py (hunk 6 above)
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: Умножение матриц в numpy является поэлементным

File: L1 - Gradient descent and linear models.py [hunk 17]
```diff
@@ -0,0 +1,701 @@
...
+plt.xlim(-7, 7)
+plt.ylim(-3, 3)
+plt.show()
+
+
+# **Градиентный спуск**
+
+# In[5]:
+
+def grad_descent(w, func, grad, learn_rate, iter_num):
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: Вообще говоря, в лабораторной не был задан критерий остановки, поэтому по логике мы были вольны выбрать любой. А фиксированное число итераций конкретно в этом случае работает довольно неплохо, так как основной задачей являлось сравнить различные методы, а отображать на графиках скорости их сходимости имеет смысл только при равном числе итераций. С помощью этого же мы можем сравнить, насколько хороший результат вышел у того или иного метода за одно и то же число итераций

File: L1 - Gradient descent and linear models.py [hunk 18]
```diff
@@ -0,0 +1,701 @@
...
+sol_mse = grad_descent(start_2d, lambda w: mse_loss(w, X, Y), lambda w: mse_grad(w, X, Y), 0.01, 300)
+yy = np.array(-(sol_mse[0][0] / sol_mse[0][1]) - (sol_mse[0][2] / sol_mse[0][1]) * x_ax)
+plot_figure_setup(hline=True, vline=True, equal_axes=True)
+plt.scatter(first.T[0], first.T[1], c='green', alpha=0.2)
+plt.scatter(second.T[0], second.T[1], c='red', alpha=0.2)
+plt.plot(x_ax, yy.T, color='b', label='Gradient descent')
+plt.plot(x_ax, y_ax_exact.T, color='black', label='Exact solution')
+plt.legend()
+plt.show()
+print('Difference between solutions norm:', np.linalg.norm(sol_mse[0] - w_exact))
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: Расстояние между направляющими векторами нормалей — вполне естественная характеристика различия; опять же, ограничения на показатели сходства прямых отсутствовали

File: L1 - Gradient descent and linear models.py [hunk 19]
```diff
@@ -0,0 +1,701 @@
...
+
+def logistic_loss(w, x, y):
+    return (1 / x.shape[0]) * sum(np.logaddexp(0, -y[i] * np.dot(x[i], w)) for i in range(x.shape[0]))
+
+
+def logistic_grad(w, x, y):
+    v = np.empty((x.shape[0], x.shape[1]))
+    for i in range(x.shape[0]):
+        val = -y[i] * np.dot(x[i], w)
+        if np.abs(val) > 1000:  # overflow protection
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: Так как image, то вместо вычисления выражения, которое вполне может уйти в inf или NaN, подставляем 1

File: L1 - Gradient descent and linear models.py [hunk 20]
```diff
@@ -0,0 +1,701 @@
...
+            v[i] = -y[i] * x[i] * (math.exp(val) / (1 + math.exp(val)))
+    return (1 / x.shape[0]) * np.sum(v, axis=0)
+
+
+# In[25]:
+
+def sgd(w, x, y, func, grad, learn_rate, pass_num, batch_size=1):
+    cur = w.copy()
+    batch_num = x.shape[0] // batch_size
+    costs = np.empty(pass_num * batch_num)
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: https://en.wikipedia.org/wiki/Loss_function
In mathematical optimization, statistics, decision theory and machine learning, a loss function or **cost** function is a function that maps an event or values of one or more variables onto a real number intuitively representing some "cost" associated with the event.

File: L1 - Gradient descent and linear models.py [hunk 21]
```diff
@@ -0,0 +1,701 @@
...
+    steps = np.empty((pass_num * batch_num, w.shape[0]))
+    for p in range(pass_num):
+        for j in range(batch_num):
+            x_new = x[j * batch_size:(j + 1) * batch_size]
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: Максимальный индекс у нас в данном случае batch_num*batch_size-1, что по построению batch_num не превысит x.shape[0]-1, то есть последний индекс массива

File: L1 - Gradient descent and linear models.py [hunk 22]
```diff
@@ -0,0 +1,701 @@
...
+            steps[p * batch_num + j] = cur
+            cur -= learn_rate * grad(cur, x_new, y_new)
+    return cur, costs, steps
+
+
+# In[26]:
+
+train = np.loadtxt('train.csv', skiprows=1, delimiter=',')
+ones = train[train[:, 0] == 1]
+zeroes = train[train[:, 0] == 0]
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: https://www.merriam-webster.com/dictionary/zero
plural zeros also zeroes

File: L1 - Gradient descent and linear models.py [hunk 23]
```diff
@@ -0,0 +1,701 @@
...
+def sgd(w, x, y, func, grad, learn_rate, pass_num, batch_size=1):
+    cur = w.copy()
+    batch_num = x.shape[0] // batch_size
+    costs = np.empty(pass_num * batch_num)
+    steps = np.empty((pass_num * batch_num, w.shape[0]))
+    for p in range(pass_num):
+        for j in range(batch_num):
+            x_new = x[j * batch_size:(j + 1) * batch_size]
+            y_new = y[j * batch_size:(j + 1) * batch_size]
+            costs[p * batch_num + j] = func(cur, x_new, y_new)
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: Здесь как раз и считается качество для всего батча. P - номер прохода по данным, j -номер батча

File: L1 - Gradient descent and linear models.py (hunk 14 above)
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: «Как влияет размер батча на скорость и качество обучения?» По оси абсцисс — размер батча, по оси ординат — качество. Или я что-то не понял, и комментарий относится к нижнему графику?

File: L1 - Gradient descent and linear models.py [hunk 24]
```diff
@@ -0,0 +1,701 @@
...
+plt.legend()
+plt.show()
+print('Difference between solutions norm:', np.linalg.norm(sol_mse[0] - w_exact))
+
+
+# **Метод наискорейшего спуска**
+
+# In[22]:
+
+def grad_descent_accel(w, func, grad, iter_num):
```
Comment (mryab) [inline L1 - Gradient descent and linear models.py]: func используется даже в стандартной библиотеке Python, grad — см. Adagrad ниже :), iter — аналогично, num достаточно общепринято

File: L1 - Gradient descent and linear models.py (hunk 21 above)
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Тогда мы не все выбираем.

File: L1 - Gradient descent and linear models.py (hunk 19 above)
Comment (dasimagin) [inline L1 - Gradient descent and linear models.py]: Виноват, не понял сразу.
